crates/protocol/src/channel_out.rs

```diff
@@ -0,0 +1,214 @@
+//! Contains the `ChannelOut` primitive for Optimism.
+
+use crate::{Batch, ChannelId, Compressor, Frame};
+use alloc::vec;
+use alloy_primitives::Bytes;
+use op_alloy_genesis::RollupConfig;
+
+/// The frame overhead.
+const FRAME_V0_OVERHEAD: usize = 23;
+
+/// An error returned by the [ChannelOut] when adding single batches.
+#[derive(Debug, Clone, PartialEq, thiserror::Error)]
+pub enum ChannelOutError {
+    /// The channel is closed.
+    #[error("The channel is already closed")]
+    ChannelClosed,
+    /// The max frame size is too small.
+    #[error("The max frame size is too small")]
+    MaxFrameSizeTooSmall,
+    /// Missing compressed batch data.
+    #[error("Missing compressed batch data")]
+    MissingData,
+    /// An error from brotli compression.
+    #[error("Error from Brotli compression")]
+    BrotliCompression,
+    /// An error encoding the `Batch`.
+    #[error("Error encoding the batch")]
+    BatchEncoding,
+    /// The encoded batch exceeds the max RLP bytes per channel.
+    #[error("The encoded batch exceeds the max RLP bytes per channel")]
+    ExceedsMaxRlpBytesPerChannel,
+}
+
+/// [ChannelOut] constructs a channel from compressed, encoded batch data.
+#[derive(Debug, Clone)]
+pub struct ChannelOut<'a, C>
+where
+    C: Compressor + Clone + core::fmt::Debug,
+{
+    /// The unique identifier for the channel.
+    pub id: ChannelId,
+    /// A reference to the [RollupConfig] used to
+    /// check the max RLP bytes per channel when
+    /// encoding and accepting the
+    pub config: &'a RollupConfig,
+    /// The rlp length of the channel.
+    pub rlp_length: u64,
+    /// Whether the channel is closed.
+    pub closed: bool,
+    /// The frame number.
+    pub frame_number: u16,
+    /// Compressed batch data.
+    pub compressed: Option<Bytes>,
+    /// The compressor.
+    pub compressor: C,
+}
+
+impl<'a, C> ChannelOut<'a, C>
+where
+    C: Compressor + Clone + core::fmt::Debug,
```

Comment:
is clone + debug required here?

Reply:
I think so in order for the struct to be `Clone + Debug` since it's composition